File: app/src/main/java/com/wuguangxin/simple/retrofit/BaseCallback.java
package com.wuguangxin.simple.retrofit;

import com.wuguangxin.utils.Logger;

import androidx.annotation.NonNull;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public abstract class BaseCallback<T extends Result<?>> implements Callback<T> {
    private static final String TAG = BaseCallback.class.getSimpleName();

    @Override
    public void onResponse(@NonNull Call<T> call, @NonNull Response<T> response) {
        int code = response.code();
        T body = response.body();
        if (code == 200) {
            onSuccess(body);
        } else {
            onError(code, response.message());
        }
        onComplete();
    }

    @Override
    public void onFailure(@NonNull Call<T> call, @NonNull Throwable t) {
        onError(-1, t.getMessage());
        onComplete();
    }

    /**
     * 当成功时回调
     */
    public void onSuccess(T response) {};

    /**
     * 当错误时回调
     */
    public void onError(int errorCode, String errorMsg) {
        Logger.e(TAG, String.format("请求失败：errorCode=%s errorMsg=%s", errorCode, errorMsg));
    }

    /**
     * 当请求完成时回调
     */
    public void onComplete() {
        Logger.i(TAG, "请求完成");
    }
}

File: app/src/main/java/com/wuguangxin/simple/retrofit/BaseCallback.kt
package com.wuguangxin.simple.retrofit

import com.wuguangxin.utils.Logger
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

abstract class BaseCallback<T : Result<*>?> : Callback<T> {

    override fun onResponse(call: Call<T>, response: Response<T>) {
        val code = response.code()
        val result = response.body()
        if (code == 200) {
            onSuccess(result)
        } else {
            onError(code, response.message())
        }
        onComplete()
    }

    override fun onFailure(call: Call<T>, t: Throwable) {
        onError(-1, t.message)
        onComplete()
    }

    /**
     * 当成功时回调
     */
    fun onSuccess(result: T?) {}

    /**
     * 当错误时回调
     */
    fun onError(code: Int, msg: String?) {
        Logger.e(TAG, "请求失败：errorCode=$code errorMsg=$msg")
    }

    /**
     * 当请求完成时回调
     */
    fun onComplete() {
        Logger.i(TAG, "请求完成")
    }

    companion object {
        private val TAG = BaseCallback::class.java.simpleName
    }
}